newIDE/app/src/MainFrame/index.js

```diff
@@ -1522,597 +1544,555 @@ class MainFrame extends React.Component<Props, State> {
       );
   };
 
-  askToCloseProject = (): Promise<void> => {
-    if (
-      this.props.unsavedChanges &&
-      this.props.unsavedChanges.hasUnsavedChanges
-    ) {
-      if (!this.state.currentProject) return Promise.resolve();
-      const { i18n } = this.props;
+  const askToCloseProject = (
+    newState: State = state,
+    newProps: Props = props
+  ): Promise<State> => {
+    const { unsavedChanges } = newProps;
+    if (unsavedChanges && unsavedChanges.hasUnsavedChanges) {
+      if (!newState.currentProject) return Promise.resolve(newState);
 
       const answer = Window.showConfirmDialog(
         i18n._(
           t`Close the project? Any changes that have not been saved will be lost.`
         )
       );
-      if (!answer) return Promise.resolve();
+      if (!answer) return Promise.resolve(newState);
     }
-    return this.closeProject();
+    return closeProject(newState);
   };
 
-  openSceneOrProjectManager = () => {
-    const { currentProject } = this.state;
+  const openSceneOrProjectManager = (newState: State = state) => {
+    const { currentProject } = newState;
     if (!currentProject) return;
 
     if (currentProject.getLayoutsCount() === 1) {
-      this.openLayout(currentProject.getLayoutAt(0).getName(), {
-        openSceneEditor: true,
-        openEventsEditor: true,
-      });
+      openLayout(
+        currentProject.getLayoutAt(0).getName(),
+        {
+          openSceneEditor: true,
+          openEventsEditor: true,
+        },
+        newState
+      );
     } else {
-      this.openProjectManager();
+      setState({ ...newState, loadingProject: false }).then(state =>
```

Comment:
Can you confirm if all the setState({ ...newState, xxx }); could be replaced by setState(state => ({...state, xxx}))?

I know it's slightly longer to write, but remove potential dependencies on "newState", that I would like to remove as much as possible.

Reply:
Okay, so basically I've passed the newState (even when it's not really needed) to break out of closure. I guess, setState(state => ({...state, xyz, ...rest})) will have the correct state and it should work fine.

I'm not sure about this though, but then again there's no harm in trying it once :)